cclient/http/http.c

```diff
@@ -0,0 +1,183 @@
+  memcpy(response->response->data + response->offset, at, length);
+  response->offset += length;
+  return RC_OK;
+}
+
+static int request_parse_message_complete(struct _response_ctx* response) {
+  response->status = IOTA_REQUEST_STATUS_DONE;
+  return RC_OK;
+}
+
+// Socket creation
+static retcode_t connect_to_iota_service(char const* const hostname,
+                                         const size_t port, int* sockfd) {
+  if ((*sockfd = open_client_socket(hostname, port)) == -1) {
+    return RC_CCLIENT_HTTP;
+  }
+  return RC_OK;
+}
+
+static retcode_t read_data_from_iota_service(int sockfd,
+                                             char_buffer_t* response) {
+  char buffer[RECEIVE_BUFFER_SIZE] = {0};
+  ssize_t num_received = 0;
+  // Setup parser settings - callbacks
+  http_parser_settings settings = {0};
+  settings.on_headers_complete = &request_parse_header_complete_cb;
+  settings.on_body = &request_parse_data_cb;
+  settings.on_message_complete = &request_parse_message_complete_cb;
+  // Setup parser
+  http_parser parser = {0};
+  // Setup response structure
+  struct _response_ctx response_context = {0};
+  response_context.parser = &parser;
+  response_context.response = response;
+  response_context.status = IOTA_REQUEST_STATUS_OK;
+  response_context.offset = 0;
+  // Initialize parser
+  http_parser_init(&parser, HTTP_RESPONSE);
+  parser.data = &response_context;
+  // Loop over received data
+  while ((num_received = receive_on_socket_wait(sockfd, buffer,
+                                                RECEIVE_BUFFER_SIZE)) > 0) {
+    int parsed = http_parser_execute(&parser, &settings, buffer, num_received);
+    // A parsing error occured, or an error in a callback
+    if (parsed < num_received ||
+        response_context.status == IOTA_REQUEST_STATUS_ERROR) {
+      return RC_CCLIENT_HTTP_RES;
+    } else if (response_context.status == IOTA_REQUEST_STATUS_DONE) {
+      return RC_OK;
+    }
+    memset(buffer, 0, sizeof(buffer));
+  }
+  if (num_received <= 0) {
+    return RC_CCLIENT_HTTP;
+  }
+  return RC_OK;
+}
+
+static retcode_t send_data_to_iota_service(int sockfd, char const* const data,
+                                           size_t data_length) {
+  char* ptr = (char*)data;
+  while (data_length > 0) {
+    ssize_t num_sent = send_on_socket_wait(sockfd, ptr, data_length);
+    if (num_sent < 0) {
+      return RC_CCLIENT_HTTP;
+    }
+    ptr += num_sent;
+    data_length -= num_sent;
+  }
+  return RC_OK;
+}
+
+static retcode_t send_headers_to_iota_service(int sockfd,
+                                              http_info_t const* http_settings,
+                                              size_t data_length) {
+  static char* header_template =
+      "POST %s HTTP/1.1\r\n"
+      "Host: %s\r\n"
+      "X-IOTA-API-Version: %d\r\n"
+      "Content-Type: application/x-www-form-urlencoded\r\n"
+      "Content-Length: %lu\r\n"
```

Comment:
The HTTP request header should specify the accepted return values (in our case `Accept: application/json\r\n`)

Reply:
Ok